### DISCUSSION TITLE
Suggestions on how to remove image colour cast

### DISCUSSION THREAD

**Discussion by kipcole9** [libvips/libvips#3319] [Help wanted]
I'd appreciate any thoughts or ideas on how to approach removing colour casts from an image.  [Fred's ImageMagick scripts](http://www.fmwconcepts.com/imagemagick/removecolorcast/index.php) has an example of the problem domain and solution but I'm pretty confident there is a more idiomatic way to approach this in `libvips` - which is currently beyond me.

**Comment by jcupitt** [comment]
Hi @kipcole9,

There are dozens of ways to do this, of course, it depends what your needs are.

The simplest thing is `scale`. This searches the image for min and max, then rescales to fit 0-255.

https://www.libvips.org/API/current/libvips-conversion.html#vips-scale

That applies the same scale and offset to every channel. You can scale the bands separately to implement "integrate to grey", the colour balance used in most photographic processes. Use `stats` to get the min and max for each band, then calculate a scale and offset for each band, then apply with `linear`.

https://www.libvips.org/API/current/libvips-arithmetic.html#vips-stats

https://www.libvips.org/API/current/libvips-arithmetic.html#vips-linear

That's doing the scale in sRGB space. You'll get a nicer result if you convert to a linear space, like scRGB, scale, then go back to sRGB. Use `colourspace` to change the colourspace.

Finally, you can get a very nice result from badly degraded images with histogram equalisation. Split it to three bands, equalise them separately, then recombine.

https://www.libvips.org/API/current/libvips-histogram.html#vips-hist-equal
